src/query/ee/src/table_lock/table_lock_handler.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright 2023 Databend Cloud
+//
+// Licensed under the Elastic License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.elastic.co/licensing/elastic-license
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+use std::time::Duration;
+
+use common_base::base::tokio::time::timeout;
+use common_base::base::GlobalInstance;
+use common_exception::ErrorCode;
+use common_exception::Result;
+use common_meta_app::schema::TableInfo;
+use common_meta_types::protobuf::watch_request::FilterType;
+use common_meta_types::protobuf::WatchRequest;
+use common_storages_fuse::TableContext;
+use common_users::UserApiProvider;
+use futures_util::StreamExt;
+use table_lock::TableLockHandler;
+use table_lock::TableLockHandlerWrapper;
+use table_lock::TableLockHeartbeat;
+
+pub struct RealTableLockHandler {}
+
+#[async_trait::async_trait]
+impl TableLockHandler for RealTableLockHandler {
+    #[async_backtrace::framed]
+    async fn try_lock(
+        &self,
+        ctx: Arc<dyn TableContext>,
+        table_info: TableInfo,
+    ) -> Result<TableLockHeartbeat> {
+        let catalog = ctx.get_catalog(table_info.catalog())?;
+        let expire_secs = ctx.get_settings().get_table_lock_expire_secs()?;
+        // get a new table lock revision.
+        let res = catalog
+            .upsert_table_lock_rev(expire_secs, &table_info, None)
+            .await?;
+        let revision = res.revision;
+
+        let table_id = table_info.ident.table_id;
+        let duration = Duration::from_secs(expire_secs);
+        let meta_api = UserApiProvider::instance().get_meta_store_client();
+        loop {
+            // List all revisions and check if the current is the minimum.
+            let reply = catalog.list_table_lock_revs(table_id).await?;
+            let position = reply.revisions.iter().position(|x| *x == revision).ok_or(
+                // If the current is not found in list,  it means that the current has expired.
+                ErrorCode::TableLockExpired("the acquired table lock has expired".to_string()),
+            )?;
+
+            if position == 0 {
+                // The lock is acquired by current session.
+                break;
+            }
+
+            // Get the previous revision, watch the delete event.
+            let key = format!("{}/{}", reply.prefix, reply.revisions[position - 1]);
```

Comment:
i think it is better to encapture a funciton such as `table_lock_key` to clarity this operation.